### DISCUSSION TITLE
What's the best pattern for externally setting the form state?

### DISCUSSION THREAD

**Discussion by r04423** [react-hook-form/react-hook-form#2886] [Q&A] [answered]
Hi!  👋

I'm having this issue where I have to synchronize form's state with URL parameters.
It's pretty easy to track form changes with `onChange` handler and then update the URL.
The harder part is to apply changes from URL to the form (for example when user goes back or forward in the browser history).

Currently I have a local state which lives outside of the form.
`useUrlParamsState` updates the URL and state when necessary to keep them in sync.
`Form` component is similar to the https://react-hook-form.com/advanced-usage#SmartFormComponent

```tsx
const [formData, setFormData] = useUrlParamsState()

<Form onChange={formData => { setFormData(formData); }}>
```

This issue is similar to the this example https://react-hook-form.com/get-started#Integratingwithglobalstate
It is possible to update global state from the form, but if some change would come from another source, this state would not be reflected in the form.

Is it possible to somehow set the form's state from outside?
If not, what's the best pattern for such use-cases?
Should I just use `useEffect` and `reset` form state on every state change? 

Thank you so much!

**Comment by bluebill1049** [comment, marked as answer]
`useEffect` with `reset` sounds good, but you probably need to find a good merging strategy, otherwise, it would confuse your users.

**Reply by r04423** [reply]
Awesome, thanks!
As for the merging strategy, do you mean merging URL and form state?

**Reply by bluebill1049** [reply]
yes

**Reply by r04423** [reply]
Yep, tried to make it as intuitive as possible 😄
Thanks again for your help and have a good one!
